src/main/java/com/google/sps/Utility.java

```diff
@@ -101,71 +106,82 @@ public static String graphToJson(
   }
 
   /**
-   * Returns the graph at the given mutation number null if the requested number is less than -1. If
-   * the user requests a number greater than the total number of mutations, we return the final
+   * Returns the graph at the given mutation number, null if the requested number is less than -1.
+   * If the user requests a number greater than the total number of mutations, we return the final
    * graph.
    *
    * @param original the original graph
    * @param curr the current (most recently-requested) graph (requires that original != curr)
-   * @param mutationNum number of mutations to apply
-   * @param multiMutList multi-mutation list
+   * @param mutationNum the index of the last mutation to apply
+   * @param multiMutList multi-mutation list builder. This parameter may be modified by replacing
+   *     some mutations with their deduplicated versions.
    * @throws IllegalArgumentException if original and current graph refer to the same object
    * @return the resulting data graph, null if the mutation number was too small, and the final
-   *     graph if the mutation number was too big
+   *     graph if the mutation number was too big.
    */
   public static DataGraph getGraphAtMutationNumber(
-      DataGraph original, DataGraph curr, int mutationNum, List<MultiMutation> multiMutList)
+      DataGraph original, DataGraph curr, int mutationNum, MutationList.Builder mutationsList)
       throws IllegalArgumentException {
     Preconditions.checkArgument(
         original != curr, "The current graph and the original graph refer to the same object");
 
+    List<MultiMutation> multiMutList = mutationsList.getMutationList();
+
     if (mutationNum < -1) {
       return null;
     } else if (mutationNum > multiMutList.size()) {
       mutationNum = multiMutList.size() - 1;
     }
+    // If the requested graph is sequentially before the current graph but is closer to
+    // the initial graph than to the current graph, go forward from the initial graph rather
+    // than going back from the current graph
+    if (curr.numMutations() - mutationNum > mutationNum) {
+      curr = original.getCopy();
+    }
     if (curr.numMutations() <= mutationNum) { // going forward
       for (int i = curr.numMutations() + 1; i <= mutationNum; i++) {
         // Mutate graph operates in place
         MultiMutation multiMut = multiMutList.get(i);
         List<Mutation> mutations = multiMut.getMutationList();
+        // Use this to store the multi mutation without any redundant mutation
+        // information (for eg. duplicate tokens to add)
+        MultiMutation.Builder trimmedMultiMut = MultiMutation.newBuilder();
```

Comment:
Right now each mutation would create a trimmedMultiMut and update the mutationsList, but only one update is needed for every mutation. One potential optimization is to keep track of which mutation is already updated to the "trimmed" version to avoid always creating a builder and updating the trimmedMultiMut.

Reply:
@EatingW could we look into it after the presentation :)